Use spf13/cobra for docker tag

yongtang opened the pull request:
This fix is part of the effort to convert commands to spf13/cobra #23211.

Thif fix coverted command `docker tag` to use spf13/cobra

Signed-off-by: Yong Tang yong.tang.github@outlook.com

yongtang committed:
Use spf13/cobra for docker tag

This fix is part of the effort to convert commands to spf13/cobra #23211.

Thif fix coverted command `docker tag` to use spf13/cobra

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

vdemeester commented on api/client/image/tag.go:
```diff
@@ -0,0 +1,38 @@
+	var opts tagOptions
+
+	cmd := &cobra.Command{
+		Use:   "tag [OPTIONS] IMAGE[:TAG] [REGISTRYHOST/][USERNAME/]NAME[:TAG]",
```
No need for `[OPTIONS]` here :angel:

yongtang replied:
Thanks @vdemeester the pull request has been updated with `[OPTIONS]` removed.

vdemeester commented on integration-cli/docker_cli_tag_test.go:
```diff
@@ -97,17 +97,17 @@ func (s *DockerSuite) TestTagWithPrefixHyphen(c *check.C) {
 	out, _, err := dockerCmdWithError("tag", "busybox:latest", "-busybox:test")
 	c.Assert(err, checker.NotNil, check.Commentf(out))
-	c.Assert(out, checker.Contains, "Error parsing reference", check.Commentf("tag a name begin with '-' should failed"))
+	c.Assert(out, checker.Contains, "unknown shorthand flag", check.Commentf("tag a name begin with '-' should failed"))
```
@dnephin hum this is a little bit weird :sweat:

yongtang replied:
Thanks @vdemeester this test has been removed and the pull request has been updated.

dnephin replied:
I suggested a fix in a comment. I don't think we should remove the tests necessarily.

dnephin commented on integration-cli/docker_cli_tag_test.go:
```diff
@@ -86,30 +86,6 @@ func (s *DockerSuite) TestTagExistedNameWithoutForce(c *check.C) {
-	out, _, err := dockerCmdWithError("tag", "busybox:latest", "-busybox:test")
-	c.Assert(err, checker.NotNil, check.Commentf(out))
-	c.Assert(out, checker.Contains, "Error parsing reference", check.Commentf("tag a name begin with '-' should failed"))
-
```
I don't think we should remove these tests. I think we can fix the behaviour.

If you check out `run`, I use `flags.SetInterspersed(false)` in the `NewRunCommand()`. This will making it stop looking for flags once it finds the first non-flag arg.

Once that' s done we should get the `-busybox:test` arg as an arg, instead of failing flag validation, which should let us get the old error message out.

yongtang replied:
Thanks @dnephin. Let me update the pull request to address this issue.

yongtang commented:
Thanks @dnephin. I added `flags.SetInterspersed(false)` though it is still shows:
```
docker tag a -b
"docker tag" requires exactly 2 argument(s).
See 'docker tag --help'.

Usage:  docker tag IMAGE[:TAG] [REGISTRYHOST/][USERNAME/]NAME[:TAG]

Tag an image into a repository
```

After some investigation, it seems that `ExactArgs()` is not passing the expected `args`.
So I changed 
```
		if len(args) == number 
```
to 
```
		if len(cmd.Flags().Args()) == number 
```
in `ExactArgs()`.

Now the command behaves as expected.

Just want to make sure there is no unintended side effect with this change.

dnephin commented:
I fixed it in cobra, and @vdemeester should be picking up the fix in one of the PRs, if you want to keep that change for now, maybe add a TODO saying it should be fixed in cobra?

yongtang commented:
@dnephin I just added a `TODO` in the commit and updated the pull request. Thanks for the help!

vdemeester commented:
Updated in #23269 :wink:

thaJeztah commented:
Oh, sorry @yongtang this one needs a rebase as well

yongtang commented:
Thanks @vdemeester I will keep an eye on #23269 and remove `TODO` if it get merged. And thanks @thaJeztah the PR has been rebased.

yongtang commented:
The pull request has been rebased. It is likely there is a failing test. The failure is caused by an issue in cobra, which has been fixed (not yet merged into the docker yet). Once #23269 is merged the the failure will be fixed.

thaJeztah commented:
@yongtang https://github.com/docker/docker/pull/23269 was merged, so you can rebase and fix the failure

yongtang commented:
Thanks @thaJeztah the PR has been rebased and fixed.

dnephin commented:
LGTM

cpuguy83 commented:
LGTM

vdemeester commented:
LGTM 🐵

thaJeztah commented:
All green! Merging
